### ISSUE TITLE
[BUG] Jobs stuck in 'SCHEDULED' state even though there are BackgroundJobServers available

### ISSUE THREAD

**Issue by nicolasmoura** [jobrunr/jobrunr#64]
**Describe the bug**
Jobs are stuck in 'SCHEDULED' state and are never picked up for processing even though there are multiple BackgroundJobServers available. Using the dashboard to "requeue" a job sends it for processing successfully.

**Environment**
I'm using version 1.2.0
I'm running on JRE / JDK 11
I'm using the following StorageProvider: MongoDBStorageProvider

**To Reproduce**
Provide a simple project or code fragment which shows how to reproduce the issue.
```java
final JobId jobId = BackgroundJob.schedule(
        () -> performJob(publicId),
        Instant.now().plus(1, ChronoUnit.MINUTES)
);
```
**Expected behavior**
I expected the job to be processed around the time it was scheduled.

**Additional context**
My app is running in a Kubernetes cluster with 4 instances up.
Storage is in MongoDB Atlas. I can confirm the app can read and write to the collections.
We use Spring Boot 2.3.5 and the following properties specified regarding JobRunr:
```yaml
org:
  jobrunr:
    background-job-server:
      enabled: true
    dashboard:
      enabled: true
```

On each instance's startup I can see the following log:
```
JobRunr BackgroundJobServer (acdb8d48-fb4c-4958-adc5-9c76592bf606) and 8 BackgroundJobPerformers started successfully
```

Thanks!

**Comment by rdehuyss** [comment]
Hi @nicolasmoura ,

Hmmm, that's really strange and should not happen off-course! 

- Can you send me the json of the jobs so I can investigate? If they are confidential, you can email them to ronald.dehuysser@gmail.com.
- In the logs it should mention that one BackgroundJobServer is master (in the same place where you see the existing log message).  Typically, this is the first server that is started. Do you see that message? Can you send me all the logs from the different servers?

Let's solve this!

**Comment by nicolasmoura** [comment]
Hi @rdehuyss, thanks for the quick reply!

I decided to do some fresh deploys so I could send you clean logs and things started working. :smiley:  

But now I think I can pinpoint the reason I was having trouble in the first place. 
There were 4 instances of my app running on Google Cloud, with the property `org.jobrunr.background-job-server.enabled` set to true. But, at the same time, there was one instance of the app running on my machine, pointing to the same database as storage provider, and also enabled as a BackgroundJobServer.

I realize now that the behavior I reported only happens when the BackgroundJobServer running on my machine is the oldest one. If I set the `background-job-server.enabled` property to false or start my instances running on cloud well before the instance running on my machine, things work smoothly and jobs are processed on schedule.

I admit that I don't know if this is expected behavior. If it is, my bad! I can close the issue. 
If it isn't, I can send you more info (job or servers JSON, logs, etc.) to help solve this. This project is awesome, would love to help!

**Comment by rdehuyss** [comment]
Well, I do find it strange as normally your laptop would then be the master and would change the state is these scheduled jobs. Is it perhaps in a different time zone?

**Comment by rdehuyss** [comment]
Hi @nicolasmoura can you confirm if your laptop was perhaps in a different timezone?

**Comment by nicolasmoura** [comment]
Hi @rdehuyss, sorry for the late reply.
The problem was not related to my computer's time zone, sadly. Having had a bit more time with it, I can also see that the issue still happens sometimes even though I disable my local instance as a BackgroundJobServer.

One thing I noticed, though, is that when my kubernetes pod containing the app instance assigned as the "master" server shuts down (gracefully), I am not finding a new "Server xxxx is master" log on my other instances. 

For example: From 0 instances running, I fresh start three instances of my app running as job servers, with ids A, B, and C. I can see the log that "Server A is master". If I then shut down instance A (master), I can't find the log saying "Server B is master" registered by the ServerZooKeeper class. Only if I shut down all instances, A, B and C, effectively shutting down all servers, and start new ones (D, E, F), that I can see the log "Server D is master" again. 

Is this expected behavior? My first guess was that when the master server was shut down, no other server was assigned master, and so other servers were not picking up jobs to execute -- but that is pure conjecture from my side haha. By the way: I can also reproduce the behavior by running multiple instances of my app locally. Can I give you any more information to help?

**Comment by rdehuyss** [comment]
Hi @nicolasmoura , it takes at least 60 seconds (and max 90 seconds) for the cluster to re-elect a new master. This is when you use default values and it depends on the `pollInterval` setting.

When the master is gone, other (so not master) `BackgroundJobServers` will continue to pickup `ENQUEUED` jobs but recurring and scheduled jobs will not move to the `ENQUEUED` state.

Is it possible you're having this issue (Jobs stuck in 'SCHEDULED' state even though there are BackgroundJobServers available) because of the fact you don't wait the 1 minute?

**Comment by nicolasmoura** [comment]
I can confirm that is not the case :confused: 
The last time that this happened (yesterday evening to morning today), the master server was shut down for at least 11 hours without a new master being re-elected. I also reproduced the scenario just now in my machine, pointing to a local MongoDB database, and it's been ~15 minutes since shutting down the master instance (no re-election so far).

Is the same log registered when a new master server is selected (`Server xxxx is master`)?

**Comment by rdehuyss** [comment]
Wauw - you are correct. I don't know how that bug snuck in. I'll keep you posted.

**Comment by rdehuyss** [comment]
Root cause found - waiting for the build to be green.

@nicolasmoura thx for spotting this - I missed an important test case.

**Comment by nicolasmoura** [comment]
@rdehuyss great! No problem, glad to be of help. Thx!

**Comment by rdehuyss** [comment]
Solved in v1.2.2

**Comment by gsabarwal-code** [comment]
Thanks was facing the same issue, upgrading the version solved the issue!
